sklearn/preprocessing/_encoders.py

```diff
@@ -1294,11 +1315,13 @@ class OrdinalEncoder(OneToOneFeatureMixin, _BaseEncoder):
         .. versionadded:: 1.3
             Read more in the :ref:`User Guide <encoder_infrequent_categories>`.
 
-    max_categories : int, default=None
+    max_categories : int or dict, default=None
```

Comment:
I'm not sure what the best way to pass the per feature maximum number of categories information is.

Using a `dict` makes sense if you only want to change the value for a few of the features. But if the input is not a data-frame then it is a bit weird to refer to columns by a name. I think it might be more natural to use the column index in that case. So something like `max_categories={2: 42}` to set the maximum number of categories to 42 for feature 2. However, looking around what other classifiers do I couldn't find a good precedent for this.

Instead we have things like the `monotonic_cst` argument for [`HistGradientBoostingClassifier`](https://scikit-learn.org/dev/modules/generated/sklearn.ensemble.HistGradientBoostingClassifier.html#sklearn.ensemble.HistGradientBoostingClassifier) and the related example: https://scikit-learn.org/dev/auto_examples/ensemble/plot_monotonic_constraints.html#monotonic-cst-features-names

Without an alternative precedent in the library I'd copy the behaviour of this kayword argument. This means `max_categories=[42, None, 256]` would specify that the max categories would be 42, `None` and 256 for features 0, 1, and 2 respectively (for the case where the input has no feature names). For data-frame input we can use a dictionary as you proposed (and as `monotonic_cst` also supports).

What do you think?

Reply:
True, I have realized that it may be unwieldy to specify the feature names in a dictionary when the input is not a data-frame, and the lack of precedent is a factor that stacks against using the feature number as a key. I think your suggestion of supporting an `array-like` for most inputs and allowing a dictionary` for a data-frame input is the best choice. In that case, passing `max_categories` as a dictionary should lead to an error if the input is not a data-frame input.